docker/scale_test.go

```diff
@@ -0,0 +1,207 @@
+package scale
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"flag"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/ethereum/go-ethereum/whisper/shhclient"
+	"github.com/ethereum/go-ethereum/whisper/whisperv5"
+	"github.com/status-im/status-go/docker/project"
+	"github.com/stretchr/testify/suite"
+)
+
+var keep = flag.Bool("keep", false, "keep the cluster after tests are finished.")
+
+type Whisp struct {
+	Name    string
+	Rpc     string
+	Metrics string
+}
+
+func MakeWhisps(containers []types.Container) []Whisp {
+	whisps := []Whisp{}
+	for _, container := range containers {
+		w := Whisp{Name: container.Names[0]}
+		for _, port := range container.Ports {
+			if port.PrivatePort == 8080 {
+				w.Metrics = fmt.Sprintf("http://%s:%d/metrics", port.IP, port.PublicPort)
+			} else if port.PrivatePort == 8545 {
+				w.Rpc = fmt.Sprintf("http://%s:%d", port.IP, port.PublicPort)
+			}
+		}
+		whisps = append(whisps, w)
+	}
+	return whisps
+}
+
+func TestWhisperScale(t *testing.T) {
+	suite.Run(t, new(WhisperScaleSuite))
+}
+
+type WhisperScaleSuite struct {
+	suite.Suite
+
+	p      project.Project
+	whisps []Whisp
+}
+
+func (w *WhisperScaleSuite) SetupSuite() {
+	flag.Parse()
+}
+
+func (s *WhisperScaleSuite) SetupTest() {
+	cli, err := client.NewEnvClient()
+	s.NoError(err)
+	s.p = project.New("wnode-test-cluster", cli)
+	s.NoError(s.p.Up(project.UpOpts{
+		Scale: map[string]int{"wnode": 12},
+		Wait:  true,
+	}))
+	containers, err := s.p.Containers(project.FilterOpts{SvcName: "wnode"})
+	s.NoError(err)
+	s.whisps = MakeWhisps(containers)
+}
+
+func (s *WhisperScaleSuite) TearDownTest() {
+	if !*keep {
+		s.NoError(s.p.Down()) // make it optional and wait
+	}
+}
+
+func (s *WhisperScaleSuite) NoErrors(errors []error) {
+	for i, err := range errors {
+		s.Require().NoError(err, "whisp", s.whisps[i], "received unexpected error")
+	}
+}
+
+func RunConcurrent(f func(i int, w Whisp) error, whisps []Whisp) []error {
+	var wg sync.WaitGroup
+	errs := make([]error, len(whisps))
+	for i, w := range whisps {
+		wg.Add(1)
+		go func(i int, w Whisp) {
+			defer wg.Done()
+			if err := f(i, w); err != nil {
+				errs[i] = err
+			}
+		}(i, w)
+	}
+	wg.Wait()
+	return errs
+}
+
+// TestSymKeyMessaging generates 100 messages with 0.5s interval between them
+// messages are generated by 5 peers
+// After all messages are generated we will wait till old new and duplicates
+// will be delivered to all peers. After that we will count how many new and old
+// envelopes received by each peer.
+func (s *WhisperScaleSuite) TestSymKeyMessaging() {
+	msgNum := 100
+	interval := 500 * time.Millisecond
+	whispCount := 5
+	if len(s.whisps) < whispCount {
+		whispCount = len(s.whisps)
+	}
+	s.NoErrors(RunConcurrent(func(i int, w Whisp) error {
+		c, err := shhclient.Dial(w.Rpc)
+		if err != nil {
+			return err
+		}
+		retries := 10
+		var info whisperv5.Info
+		for {
```

Comment:
I would extract `retry` functionality to a separate function. Now it used here and in the next `RunConcurrent` almost the same way.